content/articles/python-metaclasses-for-metaprogramming/index.md

```diff
@@ -0,0 +1,194 @@
+
+### Understanding metaclasses with examples
+Every module and function in python has a type given to it.  If the user requests to get the type of the data, a corresponding data type is returned. The `type()` function is used to determine the type of any object.
+
+```python
+number = 15  
+print("Sort of number is:", type(number))  
+   
+list = [5, 9, 10]  
+print("Sort of list is:", type(list))  
+   
+name = "John Doe"  
+print("Sort of name is:", type(name))
+```
+
+#### Output
+```python
+Sort of number is: <class 'int'>
+Sort of list is: <class 'list'>
+Sort of name is: <class 'str'>
+```
+
+Python's class system is in charge of defining all of the language's types. The above example uses class `int,` `list,` or `str`, unlike other basic data types that include `int,` `char,` and `float` as found in C and Java.
+
+New data types may be generated using user-defined classes. We will use the `Town` class as a new object in an example.
+
+```python
+class Town:  
+    pass  
+Town_object = Town()  
+print("Sort of Town_object is:", type(Town_object))
+```
+
+#### Output
+```python
+Sort of Town_object is: <class '__main__.Town'>
+```
+
+A class is an instance of Metaclass since it is an object in Python. Therefore, various class types can be found via the Metaclass, where we establish new Classes and Class objects. For instance, `town` search `type,`.
+
+```python
+class Town:  
+    pass   
+print("Sort of Town class is:", type(Town)) 
+```
+
+#### Output
+```python
+Sort of Town class is: <class 'type'>
+```
+
+We can alter the classes since they are also objects. For example, user-created properties and methods may be added or removed from the class.
+
+```python
+class Town:  
+    pass  
+Town.b = 70
+Town.foo = lambda persona: print('Dusk')
+enjoyerobject = Town()  
+   
+print(enjoyerobject.b)  
+enjoyerobject.foo() 
+```
+
+#### Output
+
+```python
+70
+Dusk
+```
+
+We may proceed by starting with a class that does not utilize any methods or variables from the class. Then, the method variables and class methods will be specified in the following steps. In the end, we will create the final object.
+
+### Reasons to use metaclasses instead of functions
+- It is apparent what the goal is.  In `UpperAttrMetaclass(type)`, it is clear what comes next.
+- OOP is an option. Metaclasses may inherit the parent method.
+- A class's subclasses are instances of its metaclass if the metaclass-class is given, but not the metaclass function.
+- You may improve the organization of your code. Using metaclasses for a simple example like this is a waste of time. It is usually reserved for a more difficult task. Making several methods and grouping them into a single class makes the code more readable.
+- It is possible to hook on `new`, `init`, and `call` hooks. So you may do other things, as a result.

```

Comment:
Explain these reasons in detail. The one-line statements provided are not clear enough.